### ISSUE TITLE
Armor slots

### ISSUE THREAD

**Issue by mmdk95** [cuberite/cuberite#2361]
I was kinda trying to implement Armors stands, but I hit a problem. I thought I could just use the inventory for the armor slots, but that is completely player dependent. After that I starting digging through the code to find an alternative in the Zombie and Skeleton classes, which should spawn on with random armor, but there were no armor slots in there either. So eventually I came to the conclusion there is no separate armor slots implementation anywhere.

So my suggestion is to split the Armor slots from the inventory and let the inventory inherit the armor slots, or the player have an armor slot instance, add the armor slots to the Zombies and Skeletons so that they can spawn with random armor on. And ass addition it should make it easier to implement the armor stands.

**Comment by worktycho** [comment]
Sounds like a good idea.

**Comment by SamJBarney** [comment]
All armor slots are are specialized inventory slots. I think the cInventory should be reworked to be independent of cPlayer, and given a variable size. That way chests, mobs, players, etc., have the same inventory management code. Then the Set and Get methods for armor just wrap a call to get/set that slot.

**Comment by mmdk95** [comment]
@SamJBarney Reworking the inventory class to suit everything that contains items? That might be even better yeah :smile:

**Comment by mmdk95** [comment]
I have thought about how this could be implemented, but I didn't really found like a "perfect" solution, but I had an idea:

There is a class called cItemGrid- which is this universal Item storage container thing. So my initial thought was to use that as the inventory and let cPawn wrap around the armor part of that, since this would enable Mobs and Players to have armor- and in the future the armor stand. The downside of this is that it would get rid of the inventory class entirely and I am worried that this might break plugins and maybe even more code.

So this was not something I really wanted to do, which lead me to my current idea- This is more or less a workaround: turn all the cPlayer pointers and what not into cPawn and in the cPawn class add a few methods:
```
// has slots?
virtual bool HasArmor()  // cPlayer does, cSkeleton and cZombie do sometimes
virtual bool HasInventory()  // cPlayer does
virtual bool HasHotBar()  //cPlayer does
```
and maybe add manipulators into cPawn too, but I dunno, I am not really happy with this way of reworking it

And here is the class structure of cPawn if that wasn't clear (with the future armor stand)
![cpawn](https://cloud.githubusercontent.com/assets/9317827/8767423/d0a6c72c-2e5c-11e5-85f0-63153d1ff92c.png)

**Comment by worktycho** [comment]
Can't you keep the old inventory if a player just has two inventories? One for armor, and one general purpose. Or am I missing something else?

**Comment by mmdk95** [comment]
@worktycho Yeah I guess that would be possible too, none of the solutions seem too great though

**Comment by worktycho** [comment]
I personally prefer the two inventories, because of the issue with players.

**Comment by mmdk95** [comment]
That might be best then

**Comment by madmaxoft** [comment]
I believe the direction we were going to use for `cPawn` descendants was to have a virtual `GetHelmet()`, `GetChestplate()` etc functions that would return the correct armor whenever valid, and an empty `cItem` when not applicable. This way it would be abstracted from any storage.

Why exactly would an armor stand be a descendant of `cPawn`? The `cPawn` class is intended to be the common ancestor of all mobs and players, there's nothing in the armor stand that would make it even a bit relevant to those. I don't see any reason for those to share any code.

**Comment by worktycho** [comment]
I don't think separate functions is a good idea. Too much shared logic.

At the moment there is an overlap as they would both have an inventory for armor. My feeling about the overlap is that they should share a sub-object, which represents an armor inventory.

**Comment by mmdk95** [comment]
I thought cause the armor stand acts like 99% like a mob and a player and the entity seemed to be a little to broad. But yeah now I look at it the armor is already represented by virtual voids in Entity, I am still learning the code ( slowly )

**Comment by SphinxC0re** [comment]
Any updates on this one?

**Comment by choryuidentify** [comment]
Any progress?

**Comment by madmaxoft** [comment]
Currently there's no manpower to handle this. We'll probably be focusing on 1.13 support first.
